index.html

```diff
@@ -1583,6 +1583,159 @@ <h5><code>writeproperty</code></h5>
+          </pre>
+          <p class="note" title="Opening a connection using JavaScript">
+            For Consumers implemented in JavaScript [[ECMASCRIPT]] and executed 
+            in a runtime which exposes the
+            <a href="https://html.spec.whatwg.org/multipage/server-sent-events.html#the-eventsource-interface">
+            <code>EventSource</code></a> interface, a Server-Sent Events 
+            connection can be initiated using the <code>EventSource</code>
+            constructor.
+            <pre class="example">
+              const levelSource = new EventSource('/things/lamp/properties/level');
+            </pre>
+          </p>
+          <p class="rfc2119-assertion" 
+            id="core-profile-protocol-binding-properties-observeproperty-5">
+            If a Web Thing receives an HTTP request following the format
+            above and the Consumer has permission to observe the 
+            corresponding property, then it MUST follow the Server-Sent Events 
+            [[EVENTSOURCE]] specification to maintain an open connection with 
+            the Consumer and push a property value to the Consumer each time 
+            the value of the specified property changes.
+          </p>
+          <p>
+            This involves the Web Thing initially sending an HTTP 
+            response to the Consumer with:
+          </p>
+          <ul>
+            <li>Status code set to <code>200</code></li>
+            <li>
+              <code>Content-Type</code> header set to 
+              <code>text/event-stream</code>
+            </li>
+          </ul>
+          <pre class="example">
+            HTTP/1.1 200 OK
+            Content-Type: text/event-stream
+          </pre>
+          <p class="rfc2119-assertion" 
+            id="core-profile-protocol-binding-properties-observeproperty-6">
+            Whenever the value of the specified property changes while the Web 
+            Thing has an open connection with a Consumer, the Web Thing MUST 
+            send a property value to the Consumer using the event stream format in
+            the Server-Sent Events [[EVENTSOURCE]] specification. For each message 
+            sent, the Web Thing MUST set the <code>event</code> field to the 
+            name of the <code>PropertyAffordance</code> and populate the 
+            <code>data</code> field with the property value, serialized in JSON 
+            and following the data schema specified in the 
+            <code>PropertyAffordance</code>. The <code>id</code> field SHOULD be set to a unique 
+            identifier for the property change, for use when re-establishing a
+            dropped connection (see below). It is RECOMMENDED that the identifier
+            is a timestamp representing the time at which the property changed, in
+            the &quotdate-time&quot format specified by [[RFC3339]].
+          </p>
+          <pre class="example">
+            event: level\n
+            data: 42\n
```

Comment:
I think two "\n" characters are needed to end the stream.

Reply:
Maybe a JSON-based serialized data would be suitable to show here, or?

Reply:
Thanks for the review.

@sebastiankb wrote:
> I think two "\n" characters are needed to end the stream.

As I understand it the double carriage return comes after the last field in the message to signify the end of the message, so it's only needed after the `id` field in this case.

> Maybe a JSON-based serialized data would be suitable to show here, or?

Technically this _is_ serialised JSON :) That might be more obvious with a property with an object type, but both the properties in the example Thing used throughout the protocol binding section are simple types (a boolean and an integer). Changing that would require changing all the examples. Alternatively we could add additional more complex examples in an appendix?

Reply:
> > Maybe a JSON-based serialized data would be suitable to show here, or?
> 
> Technically `42` _is_ serialised JSON :) That might be more obvious with a property with an object type, but both the properties in the example Thing used throughout the protocol binding section are simple types (a boolean and an integer) - see the Thing Description in #91. Changing that would require changing all the examples. Alternatively we could add additional more complex examples in an appendix?

I would encourage to think of complex properties when we design this protocol binding. Of course a single value property is trivial, and the event payload structure as well. If we delegate considerations of complex properties and cases to examples, we open the door widely for different interpretations.
For a solid specification we have to precisely define the behaviour of consumers for these cases, and I strongly support a structured event format that resolves these issues in a clear and simple way.

Please review https://github.com/w3c/wot-profile/pull/150 (again), which provides this kind of format. It is self contained specification text, i.e. does not have any dependency from other specifications, however is aligned with the CloudEvents spec to provide interop with that widely used spec.

Reply:
> I would encourage to think of complex properties when we design this protocol binding. Of course a single value property is trivial, and the event payload structure as well. If we delegate considerations of complex properties and cases to examples, we open the door widely for different interpretations.

There is no room for interpretation in the way that JSON is serialised and de-serialised from a string, that is defined in [RFC8259](https://www.rfc-editor.org/rfc/rfc8259).

An integer would be serialised in an event payload as follows:
`data: 42\n`

A more complex object would be serialised as follows:
`data: {"question":"The meaning of life, the universe and everything","answer":42}\n`

The purpose of a more complex example would simply be to more clearly illustrate that it is in fact JSON (which is already specified in an assertion), not to show how it should be serialised.

> For a solid specification we have to precisely define the behaviour of consumers for these cases, and I strongly support a structured event format that resolves these issues in a clear and simple way.

What issues need resolving exactly?

> Please review #150 (again), which provides this kind of format. It is self contained specification text, i.e. does not have any dependency from other specifications, however is aligned with the CloudEvents spec to provide interop with that widely used spec.

As already discussed in relation to event affordances (in #100, #107, #126, #134 and #150), such a data wrapper would be redundant. As far as I can tell it adds no value, all it does increase the size of the payload. What specifically do you think is missing in the payload of a property change message? What is needed except the property name, the property value and an optional timestamp?

Reply:
@benfrancis 
>Technically 42 is serialised JSON :)

That's right :-)